src/dbnode/storage/index/wide_query_results.go

```diff
@@ -0,0 +1,231 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"errors"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+// ErrWideQueryResultsExhausted is used to short circuit additional document
+// entries being added if theese wide results will no longer accept documents,
+// e.g. if the results are closed, or if no further docuemnts will pass the
+// shard filter.
+var ErrWideQueryResultsExhausted = errors.New("no more values to add to wide query results")
+
+type shardFilterFn func(ident.ID) (uint32, bool)
+
+type wideResults struct {
+	nsID   ident.ID
+	idPool ident.Pool
+
+	closed      bool
+	idsOverflow []ident.ID
+	batch       *ident.IDBatch
+	batchCh     chan<- *ident.IDBatch
+	batchSize   int
+
+	shardFilter shardFilterFn
+	shards      []uint32
+	shardIdx    int
+	// NB: pastLastShard will mark this reader as exhausted after a
+	// document is discovered whose shard exceeds the last shard this results
+	// is responsible for, using the fact that incoming documents are sorted by
+	// shard then by ID.
+	pastLastShard bool
+}
+
+// NewWideQueryResults returns a new wide query results object.
+// NB: Reader must read results from `batchCh` in a goroutine, and call
+// batch.Done() after the result is used, and the writer must close the
+// channel after no more Documents are available.
+func NewWideQueryResults(
+	namespaceID ident.ID,
+	idPool ident.Pool,
+	shardFilter shardFilterFn,
+	opts WideQueryOptions,
+) BaseResults {
+	batchSize := opts.BatchSize
+	results := &wideResults{
+		nsID:        namespaceID,
+		idPool:      idPool,
+		batchSize:   batchSize,
+		idsOverflow: make([]ident.ID, 0, batchSize),
+		batch: &ident.IDBatch{
+			IDs: make([]ident.ID, 0, batchSize),
+		},
+		batchCh: opts.IndexBatchCollector,
+		shards:  opts.ShardsQueried,
+	}
+
+	if len(opts.ShardsQueried) > 0 {
+		// Only apply filter if there are shards to filter against.
+		results.shardFilter = shardFilter
+	}
+
+	return results
+}
+
+func (r *wideResults) AddDocuments(batch []doc.Document) (int, int, error) {
```

Comment:
nit: Comment here would be helpful

Reply:
We typically don't comment methods that implement an interface, `BaseResults.AddDocuments` in this case; I think the reasoning is that it'll get shadowed anyway whenever you're using a `BaseResults`

Reply:
Ahh, my bad. I knew that, but didn't realize it was implementing an interface. SG!